src/common/tse.c

```diff
@@ -8,18 +8,166 @@
  *
  * common/tse.c
  *
- * DAOS client will use scheduler/task to manage the asynchronous tasks.
- * Tasks will be attached to one scheduler, when scheduler is executed,
- * it will walk through the task list of the scheduler and pick up those
- * ready tasks to executed.
+ * DAOS client will use scheduler/task to manage the asynchronous tasks. Tasks will be attached to
+ * one scheduler, when scheduler is executed, it will walk through the task list of the scheduler
+ * and pick up those ready tasks to executed.
  */
 #define D_LOGFAC	DD_FAC(client)
 
 #include <stdint.h>
 #include <pthread.h>
 #include <daos/common.h>
 #include <daos/tse.h>
-#include "tse_internal.h"
+
+#include <gurt/atomic.h>
+
+struct tse_task_private {
+	struct tse_sched_private *dtp_sched;
+
+	/* function for the task */
+	tse_task_func_t           dtp_func;
+
+	/* links to user task list like tse_task_list_add/_del etc APIs */
+	d_list_t                  dtp_task_list;
+
+	/* links to scheduler */
+	d_list_t                  dtp_list;
+
+	/* time to start running this task */
+	uint64_t                  dtp_wakeup_time;
+
+	/* list of tasks that depend on this task */
+	d_list_t                  dtp_dep_list;
+
+	/* daos prepare task callback list */
+	d_list_t                  dtp_prep_cb_list;
+
+	/* daos complete task callback list */
+	d_list_t                  dtp_comp_cb_list;
+
+	uint32_t
+	    /* task has been completed, no chance to be re-initialized. */
+	    dtp_completed    : 1,
+	    /* task is in running state */
+	    dtp_running      : 1,
+	    /* Don't propagate err-code from dependent tasks */
+	    dtp_no_propagate : 1, dtp_dep_cnt : 28;
+
+#if TSE_DEBUG
+	/* refcount of the task */
+	uint16_t dtp_refcnt;
+
+	uint16_t dtp_magic;
+#else
+	uint32_t dtp_refcnt;
```

Comment:
i don't like this difference in types for ref cnt for debug vs release builds.
although the ref count is never that high.

Reply:
The size of the struct is tightly controlled and @liuxuezhao wanted the full 32 bits for release builds so unless we can find a different member to adjust this is the best I can do.

I do think that 16 bits is enough for a ref count though.